barretenberg/cpp/pil/vm2/alu.pil

```diff
@@ -311,6 +286,47 @@ sel_is_u128 * sel_op_div * (1 - sel_err) * a_hi * b_hi = 0;
 #[ALU_DIV_U128]
 sel_is_u128 * sel_op_div * (1 - sel_err) * (ic * b_lo + a_lo * b_hi * TWO_POW_64 - (ia - helper1)) = 0;
 
+// FDIV
+
+sel_op_fdiv * (1 - sel_op_fdiv) = 0;
+
+// For FDIV, we reuse the same main relation for non-u128 DIV (ALU_FDIV_DIV_NON_U128) but constrain that there is no remainder, stored in helper1.
+// => The relation shows that b * c = a for FF
+
+// DIV & FDIV - shared ops
+// #[OP_ID_CHECK] ensures selectors are mutually exclusive:
+pol DIV_OPS = sel_op_div + sel_op_fdiv;
+
+pol commit sel_div_0_err;
+sel_div_0_err * (1 - sel_div_0_err) = 0;
+
+// We prove that sel_div_0_err == 1 <==> DIV_OPS == 1 && b == 0:
+//      sel_div_0_err == 1 <==> b == 0:         ib * (sel_div_0_err * (1 - b_inv) + b_inv) + sel_div_0_err - 1 = 0
+//      sel_div_0_err == 1 => DIV_OPS == 1:  sel_div_0_err * ( sel_div_0_err - DIV_OPS ) = 0
+
+// Proof:
+// ==>:
+// Assume sel_div_0_err == 1. Then #[DIV_0_ERR] becomes: DIV_OPS * ib + 1 - DIV_OPS = 0
+// We note that DIV_OPS cannot be zero, otherwise we get 1 == 0.
+// Since DIV_OPS is boolean, DIV_OPS == 1 and ib == 0
+// <== :
+// Assume DIV_OPS == 1 && ib == 0, we get for #[DIV_0_ERR]: sel_div_0_err - 1 + sel_div_0_err * (sel_div_0_err - 1) = 0
+// sel_div_0_err being boolean the second additive term is zero and therefore sel_div_0_err - 1 == 0.
+pol commit b_inv;
+#[DIV_0_ERR]
+DIV_OPS * (ib * (sel_div_0_err * (1 - b_inv) + b_inv) + sel_div_0_err - 1) + sel_div_0_err * (sel_div_0_err - DIV_OPS) = 0;
+
+// DIV & FDIV - non u128
+
+// FDIV note: I'm using (IS_NOT_U128 + sel_is_ff - IS_NOT_U128 * sel_is_ff) to avoid having to set and check sel_is_u128 for FDIV, since we know it's FF => non u128
+pol DIV_OPS_NON_U128 = DIV_OPS * (IS_NOT_U128 + sel_is_ff - IS_NOT_U128 * sel_is_ff);
+
+// Show a - remainder = b * c
+// Note: Since a, b, c, and remainder are under 64 bits (or remainder == 0 for FF), we are protected against a - remainder underflowing,
+// i.e. b * c (< 128bits) cannot equal underflowed a - remainder (> 253 bits)
+#[ALU_FDIV_DIV_NON_U128]
+DIV_OPS_NON_U128 * (1 - sel_err) * (ib * ic - (ia - sel_op_div * helper1)) = 0;
+
```

Comment:
Most of this is not new - just moved the sections of code for non-u128 `DIV` down and added cases for `FDIV`